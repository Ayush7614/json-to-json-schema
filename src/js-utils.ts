export function mapObjectValues<T, P>(
  obj: Record<string, T>,
  cb: (key: string, value: T) => P,
): Record<string, P> {
  if (obj === undefined) return undefined;
  const res = {};
  for (const key of Object.keys(obj)) {
    res[key] = cb(key, obj[key]);
  }

  return res;
}

type ObjectKey = string | number | symbol;
export function groupByField<K extends ObjectKey, TItem extends Record<K, any>>(
  collection: TItem[],
  field: K,
): { group: TItem[K]; values: TItem[] }[] {
  const groups: Record<ObjectKey, TItem[]> = {};

  for (const item of collection) {
    const group = item[field];
    groups[group] = groups[group] || [];
    groups[group].push(item);
  }

  return Object.entries(groups).map(([group, values]) => {
    return { group: group as TItem[K], values };
  });
}

export function intersect<T>(sets: Set<T>[]): Set<T> {
  if (!sets.length) return new Set();

  const i = sets.reduce((m, s, i) => (s.size < sets[m].size ? i : m), 0);
  const [smallest] = sets.splice(i, 1);

  const res = new Set<T>();

  for (let val of smallest) if (sets.every(s => s.has(val))) res.add(val);

  return res;
}
