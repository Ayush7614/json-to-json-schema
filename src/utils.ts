import { ConvertTarget, JSONSchema } from './types';

const PLAIN_OBJ_PROTOTYPE = Object.getPrototypeOf({});

export function getJsonSchemaType(value: any): string {
  if (typeof value === 'function') {
    throw new Error('Invalid JSON input, functions are not supported');
  }

  if (typeof value === 'object') {
    if (Array.isArray(value)) return 'array';
    if (value === null) return 'null';

    const proto = Object.getPrototypeOf(value);
    if (proto !== PLAIN_OBJ_PROTOTYPE && proto !== null) {
      throw new Error('Invalid JSON input');
    }

    return 'object';
  }

  if (typeof value === 'number') {
    if (Number.isInteger(value)) {
      return 'integer';
    } else {
      return 'number';
    }
  }

  return typeof value;
}

export function formatExample(example: any, target: ConvertTarget) {
  switch (target) {
    case 'draft-2020-12':
      return { examples: [example] };
    case 'draft-05-oas':
      return { example: example };
    default:
      return {};
  }
}

export function formatNull(target: ConvertTarget) {
  switch (target) {
    case 'draft-2020-12':
      return { type: 'null' };
    case 'draft-05-oas':
      return { nullable: true };
    default:
      return {};
  }
}

export function isNull(schema: JSONSchema, target: ConvertTarget) {
  switch (target) {
    case 'draft-2020-12':
      const { type } = schema;
      return type === 'null' || (Array.isArray(type) && type.some(t => t === 'null'));
    case 'draft-05-oas':
      return !!schema.nullable;
    default:
      return false;
  }
}

export function isScalarType(type: string): boolean {
  return ['string', 'number', 'integer', 'boolean', 'null'].includes(type);
}
