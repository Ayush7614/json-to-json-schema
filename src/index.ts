import { combineSchemas } from './combine';
import { detectFormat } from './format';
import { mapObjectValues } from './js-utils';
import { JSONSchema, ConvertOptions } from './types';
import { formatExample, formatNull, getJsonSchemaType } from './utils';

export function convert(example: any, options: ConvertOptions = { targetSchema: 'draft-2020-12' }): JSONSchema {
  return convertValue(example, { targetSchema: 'draft-2020-12', ...options });
}

function convertValue(value: any, options: ConvertOptions) {
  if (value === undefined) return {};

  const type = getJsonSchemaType(value);
  const exampleObj = options.includeExamples ? formatExample(value, options.targetSchema) : {};

  switch (type) {
    case 'string':
      const format = detectFormat(value);
      return { type, ...exampleObj, ...(format ? { format } : null) };
    case 'number':
    case 'integer':
    case 'boolean':
      return { type, ...exampleObj };
    case 'null':
      return formatNull(options.targetSchema);
    case 'array':
      return {
        type,
        items: value.length
          ? combineSchemas(
              value.map(v => convertValue(v, options)),
              options
            )
          : undefined,
      };
    case 'object':
      const propsLength = Object.keys(value).length;
      const additionalPropertiesObj =
        options.disableAdditionalProperties || !propsLength ? {} : { additionalProperties: false };
      return {
        type,
        properties: propsLength
          ? mapObjectValues(value, (_key, fieldValue) => {
              return convertValue(fieldValue, options);
            })
          : undefined,
        ...additionalPropertiesObj,
        // required: [],
      };
  }
}
