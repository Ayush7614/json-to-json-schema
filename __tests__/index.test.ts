import { readFileSync } from 'fs';

import { convert } from '../src';

describe('simple scalars', () => {
  const tests = [
    { name: 'undefined', example: undefined, expected: {} },
    { name: 'string', example: 'string', expected: { type: 'string' } },
    { name: 'integer', example: 1, expected: { type: 'integer' } },
    { name: 'number', example: 0.2, expected: { type: 'number' } },
    { name: 'boolean', example: false, expected: { type: 'boolean' } },
    { name: 'array', example: [], expected: { type: 'array' } },
    { name: 'object', example: {}, expected: { type: 'object' } },
    {
      name: 'null draft-05-oas',
      example: null,
      opts: { targetSchema: 'draft-05-oas' as const },
      expected: { nullable: true },
    },
    { name: 'null draft-2020-12', example: null, expected: { type: 'null' } },
  ];

  test.each(tests)('convert $name', ({ example, opts, expected }) => {
    expect(convert(example, opts)).toEqual(expected);
  });
});

describe('formats', () => {
  const tests = [
    { name: 'email', example: 'roman@redocly.com', expected: { type: 'string', format: 'email' } },
    { name: 'ipv4', example: '127.0.0.1', expected: { type: 'string', format: 'ipv4' } },
    { name: 'ipv6', example: '::1', expected: { type: 'string', format: 'ipv6' } },
    { name: 'url', example: 'https://example.com', expected: { type: 'string', format: 'url' } },
    {
      name: 'date-time',
      example: '2022-03-15T05:32:31.675Z',
      expected: { type: 'string', format: 'date-time' },
    },
    { name: 'date', example: '2022-03-15', expected: { type: 'string', format: 'date' } },
    { name: 'time', example: '05:32:31+00:00', expected: { type: 'string', format: 'time' } },
    {
      name: 'uuid',
      example: 'ca761232-ed42-11ce-bacd-00aa0057b223',
      expected: { type: 'string', format: 'uuid' },
    },
  ];

  test.each(tests)('format "$name"', ({ example, expected }) => {
    expect(convert(example)).toEqual(expected);
  });
});

describe('nested objects', () => {
  test('nested object with single scalar', () => {
    expect(convert({ a: 'test' })).toEqual({
      type: 'object',
      properties: {
        a: { type: 'string' },
      },
      additionalProperties: false,
    });
  });

  test('nested object with multiple scalar', () => {
    expect(convert({ a: 'test', b: 25 })).toEqual({
      type: 'object',
      properties: {
        a: { type: 'string' },
        b: { type: 'integer' },
      },
      additionalProperties: false,
    });
  });

  test('nested object with nested object', () => {
    expect(convert({ a: 'test', b: 25, nested: { foo: 'test@example.com' } })).toEqual({
      type: 'object',
      properties: {
        a: { type: 'string' },
        b: { type: 'integer' },
        nested: {
          type: 'object',
          properties: {
            foo: { type: 'string', format: 'email' },
          },
          additionalProperties: false,
        },
      },
      additionalProperties: false,
    });
  });
});

describe('arrays', () => {
  test('simple array of strings ', () => {
    expect(convert(['string', 'string2', 'string3'])).toEqual({
      type: 'array',
      items: {
        type: 'string',
      },
    });
  });

  test('simple array of integers ', () => {
    expect(convert([0])).toEqual({
      type: 'array',
      items: {
        type: 'integer',
      },
    });
  });

  test('simple array of strings and null ', () => {
    expect(convert(['string', null])).toEqual({
      type: 'array',
      items: {
        type: ['string', 'null'],
      },
    });

    expect(convert(['string', null], { targetSchema: 'draft-05-oas' })).toEqual({
      type: 'array',
      items: {
        type: 'string',
        nullable: true,
      },
    });
  });

  test('array of mixed scalars', () => {
    expect(convert(['test@example.com', null, 8, false])).toEqual({
      type: 'array',
      items: {
        type: ['string', 'integer', 'boolean', 'null'],
        format: 'email',
      },
    });

    expect(convert(['test@example.com', null, 8, false], { targetSchema: 'draft-05-oas' })).toEqual(
      {
        type: 'array',
        items: {
          nullable: true,
          anyOf: [{ type: 'string', format: 'email' }, { type: 'integer' }, { type: 'boolean' }],
        },
      },
    );
  });

  test('array of mixed scalars and objects', () => {
    expect(convert(['test@example.com', null, 8, false, { a: 'test' }])).toEqual({
      type: 'array',
      items: {
        anyOf: [
          { type: 'string', format: 'email' },
          { type: 'integer' },
          { type: 'boolean' },
          { type: 'object', properties: { a: { type: 'string' } }, additionalProperties: false },
          { type: 'null' },
        ],
      },
    });
  });

  test('array with nested objects', () => {
    expect(
      convert([
        {
          firstName: 'John',
          lastName: 'Doe',
          lastInvoice: null,
        },
        {
          firstName: 'James',
          lastName: 'Bond',
          lastInvoice: {
            amount: 100.01,
          },
          age: 25,
        },
      ]),
    ).toEqual({
      type: 'array',
      items: {
        type: 'object',
        properties: {
          firstName: { type: 'string' },
          lastName: { type: 'string' },
          lastInvoice: {
            type: ['object', 'null'],
            properties: {
              amount: { type: 'number' },
            },
            additionalProperties: false,
          },
          age: { type: 'integer' },
        },
        additionalProperties: false,
      },
    });
  });

  test('array with nested objects and infer required', () => {
    expect(
      convert(
        [
          {
            firstName: 'John',
            lastName: 'Doe',
            lastInvoice: null,
          },
          {
            firstName: 'James',
            lastName: 'Bond',
            lastInvoice: {
              amount: 100.01,
            },
            age: 25,
          },
        ],
        { inferRequired: true, targetSchema: 'draft-2020-12' },
      ),
    ).toEqual({
      type: 'array',
      items: {
        type: 'object',
        required: ['firstName', 'lastName', 'lastInvoice'],
        properties: {
          firstName: { type: 'string' },
          lastName: { type: 'string' },
          lastInvoice: {
            type: ['object', 'null'],
            required: ['amount'],
            properties: {
              amount: { type: 'number' },
            },
            additionalProperties: false,
          },
          age: { type: 'integer' },
        },
        additionalProperties: false,
      },
    });
  });
});

const rebillyKyc = JSON.parse(readFileSync(__dirname + '/schemas/rebilly-kyc.json', 'utf-8'));
const rebillyTransaction = JSON.parse(
  readFileSync(__dirname + '/schemas/rebilly-transaction.json', 'utf-8'),
);
const euDocument = JSON.parse(readFileSync(__dirname + '/schemas/eu-document.json', 'utf-8'));

describe('complex schemas', () => {
  test('Rebilly KYC', () => {
    expect(
      convert(rebillyKyc, {
        inferRequired: true,
        includeExamples: true,
        targetSchema: 'draft-2020-12',
      }),
    ).toMatchSnapshot();
  });

  test('Rebilly Transaction', () => {
    expect(
      convert(rebillyTransaction, {
        inferRequired: true,
        includeExamples: true,
        targetSchema: 'draft-2020-12',
      }),
    ).toMatchSnapshot();
  });

  test('EU Document', () => {
    expect(
      convert(euDocument, {
        inferRequired: true,
        includeExamples: false,
        targetSchema: 'draft-2020-12',
      }),
    ).toMatchSnapshot();
  });
});
