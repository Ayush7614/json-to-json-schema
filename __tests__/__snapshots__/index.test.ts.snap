// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`complex schemas EU Document 1`] = `
type: object
properties:
  ky:
    type: integer
  docuLanguageResource:
    type: object
    properties:
      ky:
        type: integer
      title:
        type: string
      subtitle:
        type: string
      htmlContent:
        type: string
      pdfContent:
        type: 'null'
      hasPdf:
        type: boolean
      pdfContentSize:
        type: integer
      language:
        type: string
      dolaAvResources:
        type: array
        items:
          type: object
          required:
            - ky
            - title
            - url
            - reference
            - year
            - date
            - dolaAVType
          properties:
            ky:
              type: integer
            title:
              type: string
            url:
              type: string
              format: url
            reference:
              type: string
            year:
              type: string
            date:
              type: string
              format: date
            dolaAVType:
              type: string
          additionalProperties: false
      dolaQuoteResources:
        type: array
      attachmentResources:
        type: array
      linkResources:
        type: array
    additionalProperties: false
  languages:
    type: array
    items:
      type: string
  refCd:
    type: string
  isSpeechWarning:
    type: boolean
  docutypeResource:
    type: object
    properties:
      ky:
        type: integer
      code:
        type: string
      description:
        type: string
      order:
        type: integer
    additionalProperties: false
  contactsResource:
    type: array
    items:
      type: object
      required:
        - ky
        - firstName
        - lastName
        - phone
        - mail
        - order
      properties:
        ky:
          type: integer
        firstName:
          type: string
        lastName:
          type: string
        phone:
          type: string
        mail:
          type: string
          format: email
        order:
          type: integer
      additionalProperties: false
  placeResource:
    type: object
    properties:
      ky:
        type: integer
      code:
        type: 'null'
      description:
        type: string
      order:
        type: 'null'
      parentPlace:
        type: 'null'
    additionalProperties: false
  eventDate:
    type: string
    format: date
  publishDate:
    type: string
  countriesResource:
    type: array
  policiesResource:
    type: array
    items:
      type: object
      required:
        - ky
        - code
        - description
        - order
        - piwik
        - children
      properties:
        ky:
          type: integer
        code:
          type: string
        description:
          type:
            - 'null'
        order:
          type:
            - 'null'
        piwik:
          type:
            - 'null'
        children:
          type: array
      additionalProperties: false
  isPriority:
    type: boolean
  isEmbargo:
    type: boolean
  isNoAlert:
    type: boolean
additionalProperties: false

`;

exports[`complex schemas Rebilly KYC 1`] = `
type: object
properties:
  customerId:
    type: string
    examples:
      - '123_456789'
  redirectUrl:
    type: string
    examples:
      - https://example.com/return-here/
    format: url
  documents:
    type: array
    items:
      type: object
      required:
        - type
      properties:
        type:
          type: string
          examples:
            - credit-file-proof
            - identity-proof
            - address-proof
        maxAttempts:
          type: integer
          examples:
            - 1
      additionalProperties: false
  reason:
    type: string
    examples:
      - registration
additionalProperties: false

`;

exports[`complex schemas Rebilly Transaction 1`] = `
type: object
properties:
  websiteId:
    type: string
    examples:
      - 4f6cf35x-2c4y-483z-a0a9-158621f77a21
  customerId:
    type: string
    examples:
      - 4f6cf35x-2c4y-483z-a0a9-158621f77a21
  currency:
    type: string
    examples:
      - USD
  amount:
    type: number
    examples:
      - 97.97
  invoiceIds:
    type: array
    items:
      type: string
      examples:
        - 4f6cf35x-2c4y-483z-a0a9-158621f77a21
  paymentInstruction:
    type: object
    properties:
      token:
        type: string
        examples:
          - string
    additionalProperties: false
  billingAddress:
    type: object
    properties:
      firstName:
        type: string
        examples:
          - Benjamin
      lastName:
        type: string
        examples:
          - Franklin
      organization:
        type: string
        examples:
          - Rebilly
      address:
        type: string
        examples:
          - 36 Craven St
      address2:
        type: string
        examples:
          - string
      city:
        type: string
        examples:
          - London
      region:
        type: string
        examples:
          - London
      country:
        type: string
        examples:
          - GB
      postalCode:
        type: string
        examples:
          - WC2N 5NF
      phoneNumbers:
        type: array
        items:
          type: object
          required:
            - label
            - value
            - primary
          properties:
            label:
              type: string
              examples:
                - main
            value:
              type: string
              examples:
                - 512-710-1640
            primary:
              type: boolean
              examples:
                - true
          additionalProperties: false
      emails:
        type: array
        items:
          type: object
          required:
            - label
            - value
            - primary
          properties:
            label:
              type: string
              examples:
                - main
            value:
              type: string
              format: email
              examples:
                - rebilly@example.com
            primary:
              type: boolean
              examples:
                - true
          additionalProperties: false
      dob:
        type: string
        examples:
          - '1980-04-01T00:00:00.000Z'
        format: date-time
      jobTitle:
        type: string
        examples:
          - CEO
    additionalProperties: false
  requestId:
    type: string
    examples:
      - 44433322-2c4y-483z-a0a9-158621f77a21
  gatewayAccountId:
    type: string
    examples:
      - 4f6cf35x-2c4y-483z-a0a9-158621f77a21
  description:
    type: string
    examples:
      - string
  notificationUrl:
    type: string
    examples:
      - http://example.com
    format: url
  redirectUrl:
    type: string
    examples:
      - http://example.com
    format: url
  customFields:
    type: object
    properties:
      foo:
        type: string
        examples:
          - bar
    additionalProperties: false
  riskMetadata:
    type: object
    properties:
      ipAddress:
        type: string
        examples:
          - 93.92.91.90
        format: ipv4
      fingerprint:
        type: string
        examples:
          - pIUt3xbgX3l9g3YDiLbx
      httpHeaders:
        type: object
        properties:
          Content-Type:
            type: string
            examples:
              - application/json
          Accept:
            type: string
            examples:
              - text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
        additionalProperties: false
      browserData:
        type: object
        properties:
          colorDepth:
            type: integer
            examples:
              - 24
          isJavaEnabled:
            type: boolean
            examples:
              - true
          language:
            type: string
            examples:
              - en-US
          screenWidth:
            type: integer
            examples:
              - 1920
          screenHeight:
            type: integer
            examples:
              - 1080
          timeZoneOffset:
            type: integer
            examples:
              - 300
        additionalProperties: false
      extraData:
        type: object
        properties:
          kountFraudSessionId:
            type: string
            examples:
              - abcdefg12345abababab123456789012
          payPalMerchantSessionId:
            type: string
            examples:
              - dd65ratxc5qv15iph3vyoq7l6davuowa
        additionalProperties: false
    additionalProperties: false
  isProcessedOutside:
    type: boolean
    examples:
      - false
  isMerchantInitiated:
    type: boolean
    examples:
      - false
  processedTime:
    type: string
    examples:
      - '2019-08-24T14:15:22Z'
    format: date-time
  type:
    type: string
    examples:
      - 3ds-authentication
additionalProperties: false

`;
